docs/design/hooks.md

```diff
@@ -0,0 +1,231 @@
+)
+
+// HookResource represent the resources on which hook action
+// needs to be executed
+type HookResource struct {
+	// List of Annotations, mapped to the type of resources, needs to be added on the given resource
+	Annotations map[ResourceType][]string `json:"annotations"`
+
+	// List of Finalizers, mapped to the type of resources, needs to be added on the given resource
+	Finalizers map[ResourceType][]string `json:"finalizers"`
+}
+
+// HookConfig represent the to be executed by nfs-provisioner
+type HookConfig struct {
+	// Name represent hook name
+	Name string `json:"name"`
+
+	// Resource store the resources on which this hook will be executed
+	Resource HookResource `json:"resource"`
+
+	// Event defines provisioning event on which
+	// given hook action needs to be executed
+	Event ProvisionerEventType `json:"provisioningEvent"`
+
+	// Action represent the type of hook action, i.e HookActionAdd or HookActionRemove
+	Action HookActionType `json:"hookAction"`
+}
+
+//Provisioner struct has the configuration and utilities required
+// across the different work-flows.
+type Provisioner struct {
+	stopCh chan struct{}
+
+	kubeClient clientset.Interface
+
+	// namespace in which provisioner is running
+	namespace string
+
+	// serverNamespace in which nfs server deployments gets created
+	// can be set through env variable NFS_SERVER_NAMESPACE
+	// default value is Provisioner.namespace
+	serverNamespace string
+
+	// defaultConfig is the default configurations
+	// provided from ENV or Code
+	defaultConfig []mconfig.Config
+
+	// getVolumeConfig is a reference to a function
+	getVolumeConfig GetVolumeConfigFn
+
+	//determine if clusterIP or clusterDNS should be used
+	useClusterIP bool
+
+	// k8sNodeLister hold cache information about nodes
+	k8sNodeLister listerv1.NodeLister
+
+	// nodeAffinity specifies requirements for scheduling NFS Server
+	nodeAffinity NodeAffinity
+
+    /* New Field */
+    // hooks which needs to be executed on provisioning events
+    hooks []HookConfig
+}
+```
+
+#### Configmap structure
+User needs to create Configmap resource with name **hook-config** in nfs-provisioner namespace. Hook configuration needs to be provided in data field **config**.
+Configmap needs be defined as below:
+
+```yaml
+apiVersion: v1
+data:
+  config: |
+    - hookAction: Add
+      name: hook1
+      provisioningEvent: Create
+      resource:
+        annotations:
+          BackendPV:
+          - example.io/track=true
+          - test.io/owner=teamA
+          BackendPVC:
+          - example.io/track=true
+          - test.io/owner=teamA
+          NFSPV:
+          - example.io/track=true
+          - test.io/owner=teamA
+        finalizers:
+          BackendPV:
+          - example.io/tracking-protection
+          - test.io/track=true
+          BackendPVC:
+          - example.io/tracking-protection
+          - test.io/track=true
+          NFSPV:
+          - example.io/tracking-protection
+          - test.io/track=true
```

Comment:
Have we considered something like below:
```yaml
resource:
  BackendPV:
    annotations:
    - example.io/track=true
    - test.io/owner=teamA
    finalizers:
    - example.io/tracking-protection
    - test.io/track=true
  BackendPVC:
    annotations:
    - example.io/track=true
    - test.io/owner=teamA
    finalizers:
    - example.io/tracking-protection
    - test.io/track=true
  NFSPV:
    annotations:
    - example.io/track=true
    - test.io/owner=teamA
    finalizers:
    - example.io/tracking-protection
    - test.io/track=true
  NFSServerDeployment:
    annotations:
    - example.io/track=true
    - test.io/owner=teamA
    tolerations:
    - key: openebs.io/storage
      operator: "Equal"
      value: "true"
      effect: "NoSchedule"
```
Above pattern is looking more clear and all associated configurations are available under specific resources (which helps more in programmatic perspective). While the current one is looking good in imperative perspective(`kubectl annotate <resource> name...`)